### ISSUE TITLE
Time Measurement

### ISSUE THREAD

**Issue by xyf3626** [Rendering-at-ZJU/weight-sharing-kernel-prediction-denoising#5]
Hi ! This is a great work! I want to ask you about how to measure time in this work?Like, how do you measure the predict time (5.47ms) in your paper?

**Comment by hmfann** [comment]
Hi, thanks a lot! We measure the whole denoising process in two steps: (1) the full network inference for predicting importance map; For this we measure the time simply in pytorch. (2) the filtering kernel construction from importance map and the filtering of noisy image; For this we implement a CUDA kernel (please refer to CUSTOM_WSKP_auto/CUSTOM_WSKP_auto_kernel.cu) and measure its GPU time.

**Comment by xyf3626** [comment]
> Hi, thanks a lot! We measure the whole denoising process in two steps: (1) the full network inference for predicting importance map; For this we measure the time simply in pytorch. (2) the filtering kernel construction from importance map and the filtering of noisy image; For this we implement a CUDA kernel (please refer to CUSTOM_WSKP_auto/CUSTOM_WSKP_auto_kernel.cu) and measure its GPU time.

Cool! About predict time in pytorch, the method you use is cuda.event or others? And what method do you use to measure  CUSTOM_WSKP_auto's GPU time?
More specificaly, in measurement of  predict time ,the code to measure is one or not?`outputs = model(inputs).detach()`

**Comment by hmfann** [comment]
Both of them are measured with `toch.cuda.Event` and `torch.cuda.synchronize`

**Comment by xyf3626** [comment]
> Both of them are measured with `toch.cuda.Event` and `torch.cuda.synchronize`

I use the same way to measure time, but it is too slow, the predict time I measure is 41ms and CUSTOM_WSKP_auto time is 7.49ms. I use 60 frames of classroom and my GPU is RTX 4060 Laptop which has similiar performance to RTX 2080 Ti. 
Could you please tell me why is this happened? Thank you !
This is my code below:
```
 conv_start = torch.cuda.Event(enable_timing=True)
            conv_end = torch.cuda.Event(enable_timing=True)

            conv_start.record()
            x1_out = self.lrelu1(self.conv1(x0_out))
            x2_out = self.lrelu2(self.conv2(x1_out))
            x_final_out = self.conv_final(x2_out)
            conv_end.record()
            torch.cuda.synchronize()  
            conv_time_ms = conv_start.elapsed_time(conv_end)
          
           cuda_start.record()
            x_out = CUSTOM_WSKP_auto.forward(x_irradiance, x_albedo, x_guidemap, x_alpha)

            cuda_end.record()
            torch.cuda.synchronize()  
            cuda_time_ms = cuda_start.elapsed_time(cuda_end)
```

**Comment by hmfann** [comment]
Your code seems correct. But you should fistly check if `torch.backends.cudnn.enabled` is True for convolution layer acceleration, and secondly runs the model several times in advance for GPU warmup before measurement, because GPU requires some calculations and then can be running in full power. Besides, you can run multiple times to compute the average time cost. You can try this:

```
torch.backends.cudnn.enabled = True

conv_start = torch.cuda.Event(enable_timing=True)
conv_end = torch.cuda.Event(enable_timing=True)

iterations = 100

# warmup
for i in range(iterations):
    x1_out = self.lrelu1(self.conv1(x0_out))
    x2_out = self.lrelu2(self.conv2(x1_out))
    x_final_out = self.conv_final(x2_out)

# actual measurement
conv_start.record()
for i in range(iterations):
    x1_out = self.lrelu1(self.conv1(x0_out))
    x2_out = self.lrelu2(self.conv2(x1_out))
    x_final_out = self.conv_final(x2_out)
conv_end.record()
torch.cuda.synchronize()  
conv_time_ms = conv_start.elapsed_time(conv_end) / iterations
```

**Comment by xyf3626** [comment]
> Your code seems correct. But you should fistly check if `torch.backends.cudnn.enabled` is True for convolution layer acceleration, and secondly runs the model several times in advance for GPU warmup before measurement, because GPU requires some calculations and then can be running in full power. Besides, you can run multiple times to compute the average time cost. You can try this:
> 
> ```
> torch.backends.cudnn.enabled = True
> 
> conv_start = torch.cuda.Event(enable_timing=True)
> conv_end = torch.cuda.Event(enable_timing=True)
> 
> iterations = 100
> 
> # warmup
> for i in range(iterations):
>     x1_out = self.lrelu1(self.conv1(x0_out))
>     x2_out = self.lrelu2(self.conv2(x1_out))
>     x_final_out = self.conv_final(x2_out)
> 
> # actual measurement
> conv_start.record()
> for i in range(iterations):
>     x1_out = self.lrelu1(self.conv1(x0_out))
>     x2_out = self.lrelu2(self.conv2(x1_out))
>     x_final_out = self.conv_final(x2_out)
> conv_end.record()
> torch.cuda.synchronize()  
> conv_time_ms = conv_start.elapsed_time(conv_end) / iterations
> ```

Thanks for your reply! I have already aligned with the inference time ,but I'm still confused about CUSTOM_WSKP_auto's GPU time, when I use torch.cuda.Event measure this time on RTX 3090 ,it is 1.5ms ,slower than the time 1.1ms in your paper. I want to know why did this happen?

**Comment by hmfann** [comment]
Possibly you need to warp it into a nn.Module, such as
```
class ReconstructionNet(nn.Module):
    def __init__(self):
        super(ReconstructionNet, self).__init__()
        
    def forward(self, x_irradiance, x_albedo, x_guidemap, x_alpha):
        x_out = CUSTOM_WSKP_auto.forward(x_irradiance, x_albedo, x_guidemap, x_alpha)        
        return x_out
```

**Comment by xyf3626** [comment]
Thanks! I have tried this method ,but unfortunately it didn't work.
```
cuda_start = torch.cuda.Event(enable_timing=True)
            cuda_end = torch.cuda.Event(enable_timing=True)
            re_model = ReconstructionNet()

            # warmup
            for i in range(iterations):
                x_out = re_model(x_irradiance, x_albedo, x_guidemap, x_alpha)
                # x_out = CUSTOM_WSKP_auto.forward(x_irradiance, x_albedo, x_guidemap, x_alpha)
            # 记录CUDA核函数开始
            cuda_start.record()
            for i in range(iterations):
                x_out = re_model(x_irradiance, x_albedo, x_guidemap, x_alpha)
                # x_out = CUSTOM_WSKP_auto.forward(x_irradiance, x_albedo, x_guidemap, x_alpha)
            # 记录CUDA核函数结束
            cuda_end.record()
            torch.cuda.synchronize()  # 等待事件完成
            cuda_time_ms = cuda_start.elapsed_time(cuda_end) / iterations

            x_out = x_out[:, :, self.padding_size:-self.padding_size, self.padding_size:-self.padding_size]
            print(f"Conv time: {conv_time_ms :.4f} ms, Cuda time: {cuda_time_ms :.4f} ms")
        return x_out


class ReconstructionNet(nn.Module):
    def __init__(self):
        super(ReconstructionNet, self).__init__()

    def forward(self, x_irradiance, x_albedo, x_guidemap, x_alpha):
        x_out = CUSTOM_WSKP_auto.forward(x_irradiance, x_albedo, x_guidemap, x_alpha)
        return x_out
```

**Comment by hmfann** [comment]
Possibly your profiling is correct. The cuda code was optimized in my old environment with cuda 10.x, pytorch 1.x, and old nvidia driver. Today I profiled the cuda kernel in my new environmen with cuda 12.x, pytorch 2.x, and new nvidia driver, in my 4090 GPU, the time is ~0.6ms, and in my 2080 Ti GPU, the time is ~1.8ms, so your profiling of 3090 ~1.5ms is in a reasonable range. 
I will check if it can be optimized in new CUDA versions and check if it is environment's issue in free time.

**Comment by xyf3626** [comment]
> Possibly your profiling is correct. The cuda code was optimized in my old environment with cuda 10.x, pytorch 1.x, and old nvidia driver. Today I profiled the cuda kernel in my new environmen with cuda 12.x, pytorch 2.x, and new nvidia driver, in my 4090 GPU, the time is ~0.6ms, and in my 2080 Ti GPU, the time is ~1.8ms, so your profiling of 3090 ~1.5ms is in a reasonable range. I will check if it can be optimized in new CUDA versions and check if it is environment's issue in free time.

OK. Thank you very much!

**Comment by hmfann** [comment]
You can measure the optimized cuda code below, which utilized for-loop unroll, register cache usage, and replaced the deep 'if' judgetment. In my development environment of RTX 2080ti, cuda toolkit 12.1, torch version 2.8+cu126, it runs ~0.95ms without the padding operation. By the way, since the development environment would be vary, you can report your own time mersurement in your research work if you need. Hope everything goes well!

```
#include <torch/extension.h>
#include <cuda.h>
#include <cuda_runtime.h>
#include <vector>


namespace {

// use 'constexpr' to make sure the compiler can conduct unroll loop
constexpr int C = 6;

__device__ inline float3 operator*(const float &a, const float3 &b) {
    return make_float3(a * b.x, a * b.y, a * b.z);
}

__device__ inline float3 operator*(const float3 &a, const float3 &b) {
    return make_float3(a.x * b.x, a.y * b.y, a.z * b.z);
}

__device__ inline float3 operator+(const float3 &a, const float3 &b) {
    return make_float3(a.x + b.x, a.y + b.y, a.z + b.z);
}

__device__ inline void operator+=(float3& a, const float3& b) {
    a.x += b.x;
    a.y += b.y;
    a.z += b.z;
}

template <typename scalar_t>
__global__ void CUSTOM_WSKP_auto_cuda_forward_kernel(
        const torch::PackedTensorAccessor<scalar_t, 4, torch::RestrictPtrTraits, size_t> input,
        const torch::PackedTensorAccessor<scalar_t, 4, torch::RestrictPtrTraits, size_t> albedo,
        const torch::PackedTensorAccessor<scalar_t, 4, torch::RestrictPtrTraits, size_t> guidemap,
        const torch::PackedTensorAccessor<scalar_t, 4, torch::RestrictPtrTraits, size_t> alpha,
        torch::PackedTensorAccessor<scalar_t, 4, torch::RestrictPtrTraits, size_t> output) {

    const int H = input.size(2), W = input.size(3);

    // batch index
    const int n = blockIdx.y;
    // column index
    const int c = blockIdx.x * blockDim.x + threadIdx.x;
    const int idx_h = c / W;
    const int idx_w = c % W;

    // //batch index
    // const int n = blockIdx.z;
    // // column index
    // const int idx_h = blockIdx.x * blockDim.x + threadIdx.x;
    // const int idx_w = blockIdx.y * blockDim.y + threadIdx.y;

    if (idx_h >= H || idx_w >= W) return;
    

    // register cache accumulator
    float weights_sum[C];
    float3 colors[C];

    #pragma unroll
    for (int k = 0; k < C; ++k) {
        weights_sum[k] = 0.0f;
        colors[k] = make_float3(0.0f, 0.0f, 0.0f);
    }

    float3 sub_albedo = make_float3(albedo[n][0][idx_h][idx_w], albedo[n][1][idx_h][idx_w], albedo[n][2][idx_h][idx_w]);


    // float3 sub_color, final_color = {0};
    // int h, w;
    // float weight;

    for (int i = -C; i <= C; i++) {
        const int h_in = min(max(idx_h + i, 0), H-1);

        const int abs_i = abs(i);

        for (int j = -C; j <= C; j++) {
            const int w_in = min(max(idx_w + j, 0), W-1);
                    
            const float3 sub_color = make_float3(input[n][0][h_in][w_in], input[n][1][h_in][w_in], input[n][2][h_in][w_in]);

            const int dist = max(abs_i, abs(j)); // avoid branch divergence in previous 'if'

            {
                float weight = guidemap[n][C-1][h_in][w_in];
                weights_sum[C-1] += weight;
                colors[C-1] += weight * sub_color;
            }

            const int start_k = max(1, dist); // avoid branch divergence in previous 'if'

            for (int k = C-1; k >= start_k; k--) {
                float weight = guidemap[n][k-1][h_in][w_in];
                weights_sum[k-1] += weight;
                colors[k-1] += weight * sub_color;
            }
        }
    }

    float3 final_color = make_float3(0.0f, 0.0f, 0.0f);

    #pragma unroll
    for (int k = 0; k < C; k++) {
        float w_sum = weights_sum[k];
        float alpha_val = alpha[n][k][idx_h][idx_w];
        final_color += (alpha_val / w_sum) * colors[k];
    }

    final_color = sub_albedo * final_color;
        
    output[n][0][idx_h][idx_w] = final_color.x;
    output[n][1][idx_h][idx_w] = final_color.y;
    output[n][2][idx_h][idx_w] = final_color.z;
}
    
} // namespace
    

torch::Tensor CUSTOM_WSKP_auto_cuda_forward(
        torch::Tensor input,
        torch::Tensor albedo,
        torch::Tensor guidemap,
        torch::Tensor alpha) {
    
    const auto batch_size = input.size(0);
    const auto height = input.size(2);
    const auto width = input.size(3);

    auto output = torch::zeros_like(input);

    const int threads = 512;
    // const dim3 threadss(threads, threads, 1);
    const dim3 blocks((height * width + threads - 1) / threads, batch_size);

    // const int threads = 128;
    // // const dim3 threadss(threads, threads, 1);
    // const dim3 blocks((height + threads - 1) / threads, (width + threads - 1) / threads, batch_size);


    // AT_DISPATCH_FLOATING_TYPES(input.type(), "CUSTOM_WSKP_auto_forward_cuda", ([&] {
    AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "CUSTOM_WSKP_auto_forward_cuda", ([&] {
        CUSTOM_WSKP_auto_cuda_forward_kernel<scalar_t><<<blocks, threads>>> (
            input.packed_accessor<scalar_t, 4, torch::RestrictPtrTraits, size_t>(),
            albedo.packed_accessor<scalar_t, 4, torch::RestrictPtrTraits, size_t>(),
            guidemap.packed_accessor<scalar_t, 4, torch::RestrictPtrTraits, size_t>(),
            alpha.packed_accessor<scalar_t, 4, torch::RestrictPtrTraits, size_t>(),
            output.packed_accessor<scalar_t, 4, torch::RestrictPtrTraits, size_t>());
    }));

    return output;
}
```
